modelmapper/stats.py

```diff
@@ -0,0 +1,474 @@
+import decimal
+import datetime
+from collections import Counter
+from typing import Any, NamedTuple
+
+from modelmapper.types import (
+    HasBoolean,
+    HasDateTime,
+    HasDecimal,
+    HasDollar,
+    HasInt,
+    HasNull,
+    HasPercent,
+    HasString,
+)
+
+
+class FieldStats(NamedTuple):
+    counter: Any
+    max_int: 'FieldStats' = 0
+    max_pre_decimal: 'FieldStats' = 0
+    max_decimal_scale: 'FieldStats' = 0
+    max_string_len: 'FieldStats' = 0
+    datetime_formats: 'FieldStats' = None
+    len: 'FieldStats' = 0
+
+
+class UserInferenceRequired(ValueError):
+    """Thrown when we need the prompt the user for more information"""
+    def __init__(self, value_type, message):
+        super().__init__(message)
+        self.value_type = value_type
+
+
+class NoMatchFound(ValueError):
+    """Thrown when we can't match a datatype"""
+    pass
+
+
+class InconsistentData(ValueError):
+    """Thrown when we detect inconsistent data within a field"""
+    pass
+
+
+class TypeMatcher(object):
+    """
+    Matches a value to a type
+    """
+    def match(self, item, settings):
+        return False
+
+    def is_exclusive(self):
+        """
+        Return True if a match of this type should be the _only_ reported type for a field
+        """
+        return False
+
+
+class TypeAccumulator(object):
+    """
+    Accumulates some value over every item it sees
+    """
+    def reset(self):
+        """
+        Resets the accumulator
+        """
+        pass
+
+    def inspect(self, field_name, item, settings):
+        """
+        Inspects the item
+        """
+        return False
+
+    def collect(self):
+        """
+        Sets the value of a field on the FieldStats object
+        """
+        pass
+
+
+class InSettingsMatcher(TypeMatcher):
+    """
+    Abstract Matacher -- This should be subclassed and not created directly.
+
+    Matches a value by comparing it to values in
+    the settings object.
+
+    >>> settings.test = ['1', 'y', 'true']
+    >>> InSettingsMatcher('test').match('test', settings)
+    False
+    >>> settings.test = ['1', 'y', 'true']
+    >>> InSettingsMatcher('test').match('1', settings)
+    True
+    """
+    def __init__(self, field):
+        self.field = field
+
+    def match(self, item, settings):
+        if not hasattr(settings, self.field):
+            # XXX: Raise exception
+            raise ValueError('missing field')
+            return False
```

Comment:
The `return False` wont ever run.

Reply:
Correct, it's on my TODO to decide whether or not this should throw or not, leaning towards it probably should because it's more confusing when it's not working.